bftengine/src/bftengine/PersistentStorage.hpp

```diff
@@ -0,0 +1,204 @@
+// Concord
+//
+// Copyright (c) 2018 VMware, Inc. All Rights Reserved.
+//
+// This product is licensed to you under the Apache 2.0 license (the "License").
+// You may not use this product except in compliance with the Apache 2.0
+// License.
+//
+// This product may include a number of subcomponents with separate copyright
+// notices and license terms. Your use of these subcomponents is subject to the
+// terms and conditions of the subcomponent's license, as noted in the LICENSE
+// file.
+
+#pragma once
+
+#include "PrimitiveTypes.hpp"
+#include "Bitmap.hpp"
+#include "ViewsManager.hpp"
+#include "ReplicaConfig.hpp"
+
+#include <vector>
+
+namespace bftEngine {
+namespace impl {
+class PrePrepareMsg;
+class CheckpointMsg;
+class ViewChangeMsg;
+class NewViewMsg;
+class FullCommitProofMsg;
+class PrepareFullMsg;
+class CommitFullMsg;
+
+// The PersistentStorage interface is used to write/read the concord-bft state 
+// to/from a persistent storage. In case the replica�s process is killed, 
+// crashed or restarted, this interface is used to restore the concord-bft 
+// state from the persistent storage. 
+// PersistentStorage is designed to only store data elements that are essential 
+// to reconstruct the internal in-memory data structures of concord-bft.
+// For simplicity and efficiency, some of the internal data elements are not 
+// written to the persistent storage (for example, messages with partial 
+// threshold signatures are not stored in persistent storage; if needed such 
+// messages are re-created and retransmitted by the replicas).
+
+class PersistentStorage {
```

Comment:
This pure virtual class has 44 methods. That means that every implementor has re-implement all of them to provide persistency. That seems to be a lot of effort if we want to port our persistency to different storage systems. Is there a way to change the structure of this code so that there is more abstraction and less specificity in terms of methods, perhaps by using data structs and/or templates?

For instance, the transaction methods make sense for a storage layer, but the type of data being stored is not abstracted here. I haven't gone through a full analysis, but it feels like all the getters and setters can perhaps be replaced with a single method each that take a type and return an opaque blob that can be decoded by a higher level. That decoding can be implemented once in a concrete type.

Reply:
The implementation of PersistentStorage should be based on the MetadataStorage interface. If we want to use a new storage system X, we just need to implement MetadataStorage for X.

The code that uses PersistentStorage  is relatively subtle and complex. In order to simplify it, we preferred to create a single interface with specific methods (PersistentStorage  should contain the methods that are required by the main code of “persistency and crash recovery”).